File: client-app/src/models/budgetGroup.ts
import {BudgetItem} from "./budgetItem.ts";

export class BudgetGroup {
    id: string;
    title: string;
    categories: BudgetItem[];
    assigned: number;
    outflow: number;
    available: number;
    dateCreated: string;

    constructor(id: string, name: string, dateCreated: string, categories: BudgetItem[] = [], amountAssigned: number = 0, outflow: number = 0) {
        this.id = id;
        this.title = name;
        this.categories = categories;
        this.assigned = amountAssigned;
        this.outflow = outflow;
        this.dateCreated = dateCreated;
        this.available = this.availableBalance();
    }

    public availableBalance = () => {
        return this.assigned - this.outflow;
    }
}

File: Domain/Models/Budgets/BudgetGroup.cs
namespace Domain.Models.Budgets;

public class BudgetGroup
{
    /// <summary>
    /// The <see cref="BudgetGroup.Id"/> of this <see cref="BudgetGroup"/>
    /// </summary>
    public Guid Id { get; set; }

    /// <summary>
    /// The title of this <see cref="BudgetGroup"/>
    /// </summary>
    public required string Title { get; set; }

    /// <summary>
    /// The list of <see cref="BudgetItem"/> in this <see cref="BudgetGroup"/>
    /// </summary>
    public ICollection<BudgetItem> Categories { get; set; } = new List<BudgetItem>();

    /// <summary>
    /// The date this <see cref="BudgetGroup"/> was created
    /// </summary>
    public DateTime DateCreated { get; set; }

    /// <summary>
    /// Adds a <see cref="BudgetItem"/> to the <see cref="Categories"/> list
    /// </summary>
    /// <param name="budget"></param>
    public void AddCategory(BudgetItem budget)
    {
        Categories.Add(budget);
    }
}